File: solutions/c/collatz-conjecture/1/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start) {
    if (start < 1) {
        return ERROR_VALUE;
    }
    int count = 0;
    while (start > 1) {
        ++count;
        if ((start & 1) == 0) {
            start >>= 1;
        } else {
            start += (start << 1) + 1;
        }
    }
    return count;
}

File: solutions/go/collatz-conjecture/1/collatz_conjecture.go
package collatzconjecture

import "errors"

func CollatzConjecture(n int) (steps int, e error) {
    if n <= 0 {
        e = errors.ErrUnsupported
        return
    }
    for n != 1 {
        if n % 2 == 0 {
            n /= 2
        } else {
            n = 3*n + 1
        }
        steps++
    }
    return
}
